.cbmc-batch/source/make_common_data_structures.c

```diff
@@ -13,31 +13,153 @@
+#include <proof_helpers/nondet.h>
+#include <proof_helpers/proof_allocators.h>
+#include <stdint.h>
+#include <stdlib.h>
 
-/**
- * It nondeterministically creates either a static or dynamic array list
- * structure, which is bounded by item_count and item_size parameters.
- */
-struct aws_array_list *make_arbitrary_array_list(size_t item_count, size_t item_size) {
+struct aws_array_list *make_arbitrary_array_list(size_t initial_item_allocation, size_t item_size) {
     struct aws_array_list *list;
+
     /* Assume list is allocated */
     ASSUME_VALID_MEMORY(list);
 
-    if (nondet_int()) { /* Dynamic initialization */
-        /* Use default allocator */
+    if (nondet_int()) { /* dynamic */
         struct aws_allocator *allocator = malloc(sizeof(*allocator));
-        ASSUME_DEFAULT_ALLOCATOR(allocator);
-        __CPROVER_assume(!aws_array_list_init_dynamic(list, allocator, item_count, item_size));
-    } else { /* Static initialization */
-        __CPROVER_assume(item_count > 0);
-        __CPROVER_assume(item_size > 0);
-
-        size_t len = item_count * item_size;
-        void *raw_array = malloc(len);
-
-        aws_array_list_init_static(list, raw_array, item_count, item_size);
+        ASSUME_CAN_FAIL_ALLOCATOR(allocator);
+        AWS_FATAL_ASSERT(!aws_array_list_init_dynamic(list, allocator, initial_item_allocation, item_size));
+    } else { /* static */
+        size_t len;
+        __CPROVER_assume(!aws_mul_size_checked(initial_item_allocation, item_size, &len));
+        void *raw_array = can_fail_malloc(len);
+        aws_array_list_init_static(list, raw_array, initial_item_allocation, item_size);
     }
+    list->length = nondet_size_t();
+    __CPROVER_assume(list->length <= initial_item_allocation);
+    return list;
+}
+
+struct aws_array_list *make_nondet_array_list() {
+    size_t initial_item_allocation = nondet_size_t();
+    size_t item_size = nondet_size_t();
+    struct aws_array_list *list;
+    ASSUME_ARBITRARY_ARRAY_LIST(list, initial_item_allocation, item_size);
+    return list;
+}
 
+struct aws_array_list *make_bounded_array_list(size_t max_initial_item_allocation, size_t max_item_size) {
+    size_t initial_item_allocation = nondet_size_t();
+    __CPROVER_assume(initial_item_allocation <= max_initial_item_allocation);
+    size_t item_size = nondet_size_t();
+    __CPROVER_assume(item_size <= max_item_size);
+    struct aws_array_list *list;
+    ASSUME_ARBITRARY_ARRAY_LIST(list, initial_item_allocation, item_size);
     return list;
 }
+
+int compare(const void *a, const void *b) {
+    size_t n = nondet_size_t();
+    __CPROVER_assume(n <= MAX_ITEM_SIZE);
+    __CPROVER_precondition(__CPROVER_r_ok(a, n), "first element readable in compare function");
+    __CPROVER_precondition(__CPROVER_r_ok(b, n), "second element readable in compare function");
+    return nondet_int();
+}
+
+/*
+struct aws_byte_cursor {
+    size_t len;
+    uint8_t *ptr;
+};
+*/
```

Comment:
Remove commented code

Reply:
This is not code: it is the definition of the data structure.  Having a copy of it close to the helper function makes it clearer which fields are in the structure and whether they are assigned or left nondet.